File: frontend/src/api/agents.ts
import { http, type ApiResponse } from './http'
import type { AiSource } from './ai'

export type AgentStatus =
  | 'queued'
  | 'running'
  | 'waiting_confirmation'
  | 'completed'
  | 'failed'
  | 'cancelled'

export interface PptPreferences {
  scenario: 'classroom' | 'open_lesson' | 'presentation'
  visual_style: 'serious' | 'modern' | 'youthful'
  content_density: 'concise' | 'standard' | 'detailed'
  min_slides: number
  max_slides: number
  slide_count: number
  include_interaction: boolean
  include_visuals: boolean
  template_id: number | null
}

export interface PresentationTemplate {
  id: number
  owner_id: number
  name: string
  description: string | null
  original_filename: string
  status: string
  is_shared: boolean
  slide_count: number
  aspect_ratio: string
  theme_data: {
    palette?: Record<string, string>
    fonts?: Record<string, string>
    layout_names?: string[]
    compatibility?: string
    compatibility_note?: string
  }
  created_time: string
  updated_time: string
}

export interface AgentStep {
  id: number
  step_key: string
  title: string
  step_order: number
  status: string
  output_data: Record<string, unknown>
  error_message: string | null
  started_time: string | null
  finished_time: string | null
}

export interface LessonOutline {
  title: string
  positioning: string
  objectives: {
    knowledge: string[]
    ability: string[]
    values: string[]
  }
  key_points: string[]
  difficult_points: string[]
  teaching_flow: Array<{
    stage: string
    duration_minutes: number
    teacher_activity: string
    student_activity: string
    evidence_refs: string[]
  }>
  discussion_questions: string[]
  after_class_task: string
  citation_notes: string[]
}

export interface AgentRun {
  id: number
  created_by: number
  agent_type: string
  status: AgentStatus
  course_id: number | null
  chapter_id: number | null
  teaching_class_id: number | null
  current_step: number
  input_data: Record<string, unknown>
  evidence_snapshot: AiSource[]
  output_data: {
    outline?: LessonOutline
    artifact_bundle?: {
      ppt?: {
        title: string
        subtitle?: string
        design?: {
          name: string
          concept: string
          mood: string
          status: 'personalized' | 'fallback'
          designed_pages: number
          palette: Record<string, string>
          fonts?: Record<string, string>
          template_reference?: {
            id: number
            name: string
            compatibility?: string
          }
        }
        quality_report?: {
          score: number
          passed: boolean
          summary: string
          reviewer: string
          issues: Array<{
            slide_index: number | null
            severity: 'high' | 'medium' | 'low'
            category: string
            message: string
            suggestion: string
          }>
        }
        slides: Array<{
          layout: string
          title: string
          takeaway: string
          bullets: string[]
          keyword?: string
          left?: {
            title: string
            points: string[]
          }
          right?: {
            title: string
            points: string[]
          }
          steps?: Array<{
            title: string
            description: string
          }>
          timeline?: Array<{
            label: string
            title: string
          }>
          canvas_background?: string
          canvas?: Array<{
            type: 'text' | 'shape' | 'line' | 'image'
            source: string
            style: 'hero' | 'title' | 'subtitle' | 'body' | 'label' | 'number' | 'quote'
            x: number
            y: number
            w: number
            h: number
            color: string
            fill?: string
            shape: 'rect' | 'roundRect' | 'ellipse' | 'arc'
            align: 'left' | 'center' | 'right'
            bold: boolean
          }>
          visual_asset?: {
            file_name: string
            media_type: string
            model: string
          }
          speaker_notes: string
          evidence_refs: string[]
        }>
        multimodal?: {
          status: 'completed' | 'fallback' | 'unavailable'
          generated_count: number
          requested_count?: number
          model?: string
          message: string
          errors?: string[]
        }
      }
      lesson_plan?: {
        title: string
        overview: string
        objectives: string[]
        preparation: string[]
        procedures: LessonOutline['teaching_flow']
        assessment: string[]
        homework: string
      }
      classroom_activities?: Array<{
        title: string
        purpose: string
        duration_minutes: number
        format: string
        instructions: string[]
        questions: string[]
        evidence_refs: string[]
        evaluation: string
      }>
    }
    artifacts?: Record<string, {
      kind: 'ppt' | 'lesson_plan' | 'classroom_activities'
      title: string
      file_name: string
      media_type: string
      slide_count?: number
      activity_count?: number
      preview: unknown
    }>
    ppt_versions?: Array<{
      version_id: string
      created_time: string
      reason: string
      title: string
      slide_count: number
      design_name?: string
    }>
  }
  model_name: string | null
  prompt_version: string
  error_message: string | null
  cancel_requested: boolean
  retry_of_run_id: number | null
  started_time: string | null
  finished_time: string | null
  created_time: string
  updated_time: string
  steps: AgentStep[]
}

export interface CreateLessonPrepRun {
  agent_type: 'teacher_lesson_prep'
  course_id: number
  chapter_id: number
  teaching_class_id: number | null
  input: {
    lesson_hours: number
    student_level: string
    teaching_goal: string | null
    output_types: Array<'outline' | 'lesson_plan' | 'ppt'>
  }
}

export interface AgentCapabilities {
  ppt_multimodal_available: boolean
  ppt_multimodal_model: string | null
  ppt_multimodal_max_images: number
}

export interface LessonPublication {
  id: number
  agent_run_id: number
  teaching_class_id: number
  teaching_class_name: string
  course_id: number
  chapter_id: number
  chapter_title: string
  created_by: number
  teacher_name: string
  title: string
  description: string
  ppt_available: boolean
  ppt_file_name: string | null
  discussion_activity_ids: number[]
  status: string
  created_time: string
}

export const agentApi = {
  capabilities: () =>
    http.get<ApiResponse<AgentCapabilities>>('/agent/capabilities'),
  create: (payload: CreateLessonPrepRun) =>
    http.post<ApiResponse<AgentRun>>('/agent/runs', payload, { timeout: 90_000 }),
  list: (limit = 30) =>
    http.get<ApiResponse<AgentRun[]>>('/agent/runs', { params: { limit } }),
  detail: (id: number) =>
    http.get<ApiResponse<AgentRun>>(`/agent/runs/${id}`),
  confirmEvidence: (id: number) =>
    http.post<ApiResponse<AgentRun>>(`/agent/runs/${id}/confirm`, { action: 'approve_evidence' }),
  generateArtifacts: (
    id: number,
    outputTypes: Array<'lesson_plan' | 'ppt' | 'classroom_activities'>,
    pptPreferences?: PptPreferences,
  ) =>
    http.post<ApiResponse<AgentRun>>(`/agent/runs/${id}/artifacts`, {
      output_types: outputTypes,
      ppt_preferences: outputTypes.includes('ppt') ? pptPreferences : undefined,
    }),
  revisePptSlide: (
    id: number,
    slideIndex: number,
    instruction: string,
    mode: 'content' | 'design' | 'both',
  ) =>
    http.post<ApiResponse<AgentRun>>(`/agent/runs/${id}/ppt/slides/${slideIndex}/revise`, {
      instruction,
      mode,
    }, { timeout: 180_000 }),
  restorePptVersion: (id: number, versionId: string) =>
    http.post<ApiResponse<AgentRun>>(`/agent/runs/${id}/ppt/versions/restore`, {
      version_id: versionId,
    }, { timeout: 120_000 }),
  listPptTemplates: () =>
    http.get<ApiResponse<PresentationTemplate[]>>('/agent/ppt-templates'),
  uploadPptTemplate: (form: FormData) =>
    http.post<ApiResponse<PresentationTemplate>>('/agent/ppt-templates', form, {
      timeout: 120_000,
    }),
  deletePptTemplate: (id: number) =>
    http.delete<ApiResponse<null>>(`/agent/ppt-templates/${id}`),
  publish: (id: number, payload: {
    teaching_class_id: number
    title: string
    description: string
    publish_ppt: boolean
    publish_discussions: boolean
    discussion_indices: number[]
    confirmed: boolean
  }) =>
    http.post<ApiResponse<LessonPublication>>(`/agent/runs/${id}/publish`, payload),
  publications: (teachingClassId?: number) =>
    http.get<ApiResponse<LessonPublication[]>>('/agent/publications', {
      params: teachingClassId ? { teaching_class_id: teachingClassId } : undefined,
    }),
  downloadPublicationPpt: (id: number) =>
    http.get<Blob>(`/agent/publications/${id}/ppt`, { responseType: 'blob' }),
  downloadArtifact: (id: number, artifactKey: string) =>
    http.get<Blob>(`/agent/runs/${id}/artifacts/${artifactKey}/download`, {
      responseType: 'blob',
    }),
  cancel: (id: number) =>
    http.post<ApiResponse<AgentRun>>(`/agent/runs/${id}/cancel`),
  retry: (id: number) =>
    http.post<ApiResponse<AgentRun>>(`/agent/runs/${id}/retry`),
  async stream(id: number, handlers: {
    onSnapshot: (run: AgentRun) => void
    onDone?: () => void
  }) {
    const baseURL = import.meta.env.VITE_API_BASE_URL || '/api/v1'
    const response = await fetch(`${baseURL}/agent/runs/${id}/events`, {
      headers: { Authorization: `Bearer ${localStorage.getItem('access_token') || ''}` },
    })
    if (!response.ok || !response.body) throw new Error('智能任务进度服务暂时不可用')
    const reader = response.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''
    while (true) {
      const { value, done } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const events = buffer.split('\n\n')
      buffer = events.pop() || ''
      for (const block of events) {
        const event = block.match(/^event:\s*(.+)$/m)?.[1]
        const dataLine = block.match(/^data:\s*(.+)$/m)?.[1]
        if (!event || !dataLine) continue
        const data = JSON.parse(dataLine)
        if (event === 'snapshot') handlers.onSnapshot(data)
        if (event === 'done') handlers.onDone?.()
        if (event === 'error') throw new Error(data.message || '智能任务执行失败')
      }
    }
  },
}

File: backend/app/api/v1/endpoints/agents.py
import json
from time import sleep
from urllib.parse import quote

from fastapi import APIRouter, BackgroundTasks, Depends, File, Form, Query, UploadFile
from fastapi.responses import FileResponse, StreamingResponse
from sqlalchemy import Engine
from sqlalchemy.orm import Session

from app.api.dependencies import require_roles
from app.core.config import settings
from app.db.session import get_db
from app.models.agent_run import AgentRun
from app.models.user import User
from app.schemas.agent import (
    AgentArtifactRequest,
    AgentConfirmRequest,
    AgentRunCreate,
    AgentRunData,
    AgentCapabilities,
    LessonPublicationData,
    LessonPublishRequest,
    PptSlideRevisionRequest,
    PptVersionRestoreRequest,
    PresentationTemplateData,
)
from app.schemas.common import ApiResponse, api_json_value
from app.services.agent_service import (
    AgentService,
    execute_lesson_artifacts,
    execute_lesson_outline,
)
from app.services.ai_operation_service import AiProviderConfigService
from app.services.presentation_template_service import PresentationTemplateService
from app.services.lesson_publication_service import LessonPublicationService


router = APIRouter(prefix="/agent", tags=["agent"])


@router.get("/capabilities", response_model=ApiResponse[AgentCapabilities])
def agent_capabilities(
    _: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentCapabilities]:
    config = AiProviderConfigService.resolve_capability("image_generation", db)
    available = bool(
        config.enabled
        and config.api_key
        and config.base_url
        and config.model_name
    )
    return ApiResponse(
        data=AgentCapabilities(
            ppt_multimodal_available=available,
            ppt_multimodal_model=config.model_name if available else None,
            ppt_multimodal_max_images=settings.ppt_multimodal_max_images if available else 0,
        )
    )


@router.get("/ppt-templates", response_model=ApiResponse[list[PresentationTemplateData]])
def list_ppt_templates(
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[list[PresentationTemplateData]]:
    return ApiResponse(data=PresentationTemplateService(db, user).list())


@router.post("/ppt-templates", response_model=ApiResponse[PresentationTemplateData])
async def upload_ppt_template(
    name: str = Form(..., min_length=1, max_length=120),
    description: str | None = Form(default=None, max_length=1000),
    is_shared: bool = Form(default=False),
    file: UploadFile = File(...),
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[PresentationTemplateData]:
    content = await file.read()
    data = PresentationTemplateService(db, user).create(
        name=name,
        description=description,
        is_shared=is_shared,
        original_filename=file.filename or "presentation-template.pptx",
        content=content,
    )
    return ApiResponse(message="PPT 模板已解析，可在生成设置中选择", data=data)


@router.delete("/ppt-templates/{template_id}", response_model=ApiResponse[None])
def delete_ppt_template(
    template_id: int,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[None]:
    PresentationTemplateService(db, user).delete(template_id)
    return ApiResponse(message="PPT 模板已删除")


@router.get("/publications", response_model=ApiResponse[list[LessonPublicationData]])
def list_lesson_publications(
    teaching_class_id: int | None = Query(default=None, ge=1),
    user: User = Depends(require_roles("student", "teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[list[LessonPublicationData]]:
    return ApiResponse(data=LessonPublicationService(db, user).list(teaching_class_id))


@router.get("/publications/{publication_id}/ppt")
def download_published_ppt(
    publication_id: int,
    user: User = Depends(require_roles("student", "teacher", "admin")),
    db: Session = Depends(get_db),
) -> FileResponse:
    path, item = LessonPublicationService(db, user).download(publication_id)
    filename = item.ppt_file_name or path.name
    response = FileResponse(
        path,
        media_type="application/vnd.openxmlformats-officedocument.presentationml.presentation",
        filename=filename,
    )
    response.headers["Content-Disposition"] = (
        f"attachment; filename*=UTF-8''{quote(str(filename))}"
    )
    return response


@router.post("/runs", response_model=ApiResponse[AgentRunData])
def create_run(
    payload: AgentRunCreate,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    data = AgentService(db, user).create(payload)
    return ApiResponse(message="证据包已构建，请人工确认后生成课纲", data=data)


@router.get("/runs", response_model=ApiResponse[list[AgentRunData]])
def list_runs(
    limit: int = Query(default=30, ge=1, le=100),
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[list[AgentRunData]]:
    return ApiResponse(data=AgentService(db, user).list(limit))


@router.get("/runs/{run_id}", response_model=ApiResponse[AgentRunData])
def get_run(
    run_id: int,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    return ApiResponse(data=AgentService(db, user).get(run_id))


@router.post("/runs/{run_id}/confirm", response_model=ApiResponse[AgentRunData])
def confirm_run(
    run_id: int,
    payload: AgentConfirmRequest,
    background_tasks: BackgroundTasks,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    service = AgentService(db, user)
    if payload.action == "approve_evidence":
        data = service.approve_evidence(run_id)
        background_tasks.add_task(execute_lesson_outline, run_id, db.get_bind())
        return ApiResponse(message="证据已确认，课纲正在后台生成", data=data)
    return ApiResponse(message="确认完成", data=service.get(run_id))


@router.post("/runs/{run_id}/artifacts", response_model=ApiResponse[AgentRunData])
def generate_artifacts(
    run_id: int,
    payload: AgentArtifactRequest,
    background_tasks: BackgroundTasks,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    data = AgentService(db, user).request_artifacts(run_id, payload)
    background_tasks.add_task(execute_lesson_artifacts, run_id, db.get_bind())
    return ApiResponse(message="教学成果已进入后台生成", data=data)


@router.post(
    "/runs/{run_id}/publish",
    response_model=ApiResponse[LessonPublicationData],
)
def publish_lesson_artifacts(
    run_id: int,
    payload: LessonPublishRequest,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[LessonPublicationData]:
    data = LessonPublicationService(db, user).publish(run_id, payload)
    return ApiResponse(message="PPT 与课堂讨论已发布到教学班", data=data)


@router.post(
    "/runs/{run_id}/ppt/slides/{slide_index}/revise",
    response_model=ApiResponse[AgentRunData],
)
def revise_ppt_slide(
    run_id: int,
    slide_index: int,
    payload: PptSlideRevisionRequest,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    data = AgentService(db, user).revise_ppt_slide(run_id, slide_index, payload)
    return ApiResponse(message="本页已重新生成，并保留上一版本", data=data)


@router.post("/runs/{run_id}/ppt/versions/restore", response_model=ApiResponse[AgentRunData])
def restore_ppt_version(
    run_id: int,
    payload: PptVersionRestoreRequest,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    data = AgentService(db, user).restore_ppt_version(run_id, payload.version_id)
    return ApiResponse(message="已恢复所选 PPT 版本", data=data)


@router.get("/runs/{run_id}/artifacts/{artifact_key}/download")
def download_artifact(
    run_id: int,
    artifact_key: str,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> FileResponse:
    path, artifact = AgentService(db, user).artifact_download(run_id, artifact_key)
    filename = artifact.get("file_name") or path.name
    response = FileResponse(
        path,
        media_type=artifact.get("media_type") or "application/octet-stream",
        filename=filename,
    )
    response.headers["Content-Disposition"] = (
        f"attachment; filename*=UTF-8''{quote(str(filename))}"
    )
    return response


@router.post("/runs/{run_id}/cancel", response_model=ApiResponse[AgentRunData])
def cancel_run(
    run_id: int,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    return ApiResponse(message="任务已取消", data=AgentService(db, user).cancel(run_id))


@router.post("/runs/{run_id}/retry", response_model=ApiResponse[AgentRunData])
def retry_run(
    run_id: int,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> ApiResponse[AgentRunData]:
    return ApiResponse(message="已创建重试任务", data=AgentService(db, user).retry(run_id))


@router.get("/runs/{run_id}/events")
def run_events(
    run_id: int,
    user: User = Depends(require_roles("teacher", "admin")),
    db: Session = Depends(get_db),
) -> StreamingResponse:
    AgentService(db, user).get(run_id)
    bind = db.get_bind()
    user_id = user.id
    is_admin = user.role == "admin"

    def event_stream():
        previous = ""
        while True:
            with Session(bind=bind, autoflush=False, expire_on_commit=False) as event_db:
                run = event_db.get(AgentRun, run_id)
                if run is None or (not is_admin and run.created_by != user_id):
                    yield f"event: error\ndata: {json.dumps({'message': '任务不存在或无权访问'}, ensure_ascii=False)}\n\n"
                    return
                data = api_json_value(AgentService.serialize(event_db, run))
                serialized = json.dumps(data, ensure_ascii=False)
                if serialized != previous:
                    yield f"event: snapshot\ndata: {serialized}\n\n"
                    previous = serialized
                if run.status in {"waiting_confirmation", "completed", "failed", "cancelled"}:
                    yield "event: done\ndata: {}\n\n"
                    return
            sleep(0.5)

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )
